pkg/compactor/block_upload.go

```diff
@@ -420,3 +530,48 @@ func (r bodyReader) ObjectSize() (int64, error) {
 func (r bodyReader) Read(b []byte) (int, error) {
 	return r.r.Body.Read(b)
 }
+
+type ValidationProgress struct {
+	Timestamp   int64  // current time
+	ObjectCount int64  // objects downloaded
+	ObjectBytes int64  // total bytes downloaded
+	Finished    bool   // will be set to true for last message
+	Success     bool   // only when finished = true
+	Message     string // message for the user.
+}
+
+type requestProgressSender struct {
+	resp http.ResponseWriter
+	enc  *json.Encoder
+}
+
+func newRequestProgressSender(w http.ResponseWriter) *requestProgressSender {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+
+	return &requestProgressSender{
+		resp: w,
+		enc:  json.NewEncoder(w),
+	}
+}
+
+func (rps *requestProgressSender) SendValidationProgress(p ValidationProgress) error {
+	return rps.enc.Encode(p)
+}
```

Comment:
Let's make this private, and use it from other methods. Let's also set timestamp at single place (ie. in this method).